### PR TITLE
Fixed multi-epass2003 problem and improve code

### PR DESCRIPTION
**Pull request by FeitianSmartcardReader**
Fixed bug #731 #730

### CODE REVIEW

File: src/libopensc/card-epass2003.c [hunk 1]
```diff
@@ -324,10 +328,10 @@ gen_init_key(struct sc_card *card, unsigned char *key_enc, unsigned char *key_ma
 	apdu.le = apdu.resplen = 28;
 	apdu.resp = result;	/* card random is result[12~19] */
 
-	tmp_sm = g_sm;
-	g_sm = SM_PLAIN;
+	tmp_sm = exdata->sm;
```
Comment (frankmorgner) [inline src/libopensc/card-epass2003.c]: You could check if `exdata != NULL`

Comment (frankmorgner) [inline src/libopensc/card-epass2003.c]: This should always be done before dereferencing exdata

Comment (frankmorgner) [inline src/libopensc/card-epass2003.c]: (I think there should not be any problems. Its just a precautionary measure.)

**Comment by frankmorgner** [comment]
@velter could you have a look?

**Comment by velter** [comment]
@frankmorgner This look like to be the same patch as #731 but applied to master instead of towards-opensc-0.16.0

I agree with the missing dereferencing test. What do you think about adding

LOG_TEST_RET(card->ctx, exdata?0:SC_ERROR_INVALID_ARGUMENTS, "unable to get extended driver data");

in each function trying to access card->drv_data ?

I can provide a patch if you are ok with this.

**Comment by frankmorgner** [comment]
If you could provide a PR, that would be great. I'd use a simple
```
if (!card->drv_data) return SC_ERROR_INVALID_ARGUMENTS
```

**Comment by FeitianSmartcardReader** [comment]
@frankmorgner and @velter  modify the code, have a look, thanks

**Comment by velter** [comment]
@frankmorgner PR #780 add check for invalid driver datas. The PR is against the towards-opensc-0.16.0 branch

**Comment by FeitianSmartcardReader** [comment]
@frankmorgner just modify card-entersafe.c file, add new ATR support and also double check the source code, add code to avoid memory leak, thanks

**Comment by frankmorgner** [comment]
before adding a commit, please at least try to compile it:
```
card-epass2003.c(324): error C2146: syntax error: missing ';' before identifier 'epass2003_exdata'
```

**Comment by frankmorgner** [comment]
@FeitianSmartcardReader I am expecting you to maintain the code of your cards.

**Comment by FeitianSmartcardReader** [comment]
@frankmorgner  Sorry for that. Did fix it and compile it to make sure.
Have a check, thanks https://github.com/OpenSC/OpenSC/pull/770/commits/7502c2727e55b841648f9df84297f83ca476af0d

**Comment by frankmorgner** [comment]
Could you please squash 7502c2727e55b841648f9df84297f83ca476af0d onto a3dcc0cd820f411c6f5ae8de0d5988545f8825ec so that it appears as one single commit? Thanks!

**Comment by FeitianSmartcardReader** [comment]
Thanks @frankmorgner , Will submit another pull request to make clearly report in OpenSC.

**Pull request closed without merging**